File: go/nucleotide-count/nucleotide_count.go
package dna

import "errors"

// Histogram is a mapping from nucleotide to its count in given DNA.
type Histogram map[rune]int

// DNA is a list of nucleotides.
type DNA string

// Counts generates a histogram of valid nucleotides in the given DNA.
func (d DNA) Counts() (Histogram, error) {
    h := Histogram{'A': 0, 'C': 0, 'G': 0, 'T': 0}
    for _, char := range d {
        if value, ok := h[char]; ok {
        	h[char] = value + 1
        } else {
        	var emptyHist Histogram
        	return emptyHist, errors.New("input contains invalid character")
        }
    }
	return h, nil
}

File: cpp/nucleotide-count/nucleotide_count.cpp
#include "nucleotide_count.h"
#include <map>
#include <stdexcept>

namespace nucleotide_count {
    std::map<char, int> count(std::string strand) {
        std::map<char, int> counts {{'A', 0}, {'C', 0}, {'G', 0}, {'T', 0}};
        for (char c: strand) {
            counts[c]++;
            if (counts.size() > 4) throw std::invalid_argument("invalid nucleotide");
        }
        return counts;
    }
}  // namespace nucleotide_count
